File: android/src/main/java/com/getcapacitor/community/firebasecrashlytics/FirebaseCrashlyticsPlugin.kt
package com.getcapacitor.community.firebasecrashlytics

import android.Manifest
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginException
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission

@CapacitorPlugin(
    name = "FirebaseCrashlytics",
    permissions = [
        Permission(strings = [Manifest.permission.ACCESS_NETWORK_STATE], alias = "network"),
        Permission(strings = [Manifest.permission.INTERNET], alias = "internet"),
        Permission(strings = [Manifest.permission.WAKE_LOCK], alias = "wakelock")
    ]
)
public class FirebaseCrashlyticsPlugin : Plugin() {
    private lateinit var implementation: FirebaseCrashlytics

    override fun load() {
        implementation = FirebaseCrashlytics()
    }

    @PluginMethod
    public fun crash(call: PluginCall) {
        val message = call.getString("message") ?: throw PluginException(ERROR_MESSAGE_MISSING)
        call.resolve()
        implementation.crash(message)
    }

    @PluginMethod
    public fun setContext(call: PluginCall) {
        val key = call.getString("key") ?: throw PluginException(ERROR_KEY_MISSING)
        val hasValue = call.data.has("value")
        if (!hasValue) {
            throw PluginException(ERROR_VALUE_MISSING)
        }
        // "string" is the default, so type is never null here
        val type = call.getString("type", "string")!!
        implementation.setContext(key, type, call)
        call.resolve()
    }

    @PluginMethod
    public fun setUserId(call: PluginCall) {
        val userId = call.getString("userId") ?: throw PluginException(ERROR_USERID_MISSING)
        implementation.setUserId(userId)
        call.resolve()
    }

    @PluginMethod
    public fun addLogMessage(call: PluginCall) {
        val message = call.getString("message") ?: throw PluginException(ERROR_MESSAGE_MISSING)
        implementation.addLogMessage(message)
        call.resolve()
    }

    @PluginMethod
    public fun setEnabled(call: PluginCall) {
        val enabled = call.getBoolean("enabled") ?: throw PluginException(ERROR_ENABLED_MISSING)
        implementation.setEnabled(enabled)
        call.resolve()
    }

    @PluginMethod
    public fun isEnabled(call: PluginCall) {
        call.unimplemented("Not implemented on Android.")
    }

    @PluginMethod
    public fun didCrashDuringPreviousExecution(call: PluginCall) {
        val crashed = implementation.didCrashDuringPreviousExecution()
        val ret = JSObject()
        ret.put("crashed", crashed)
        call.resolve(ret)
    }

    @PluginMethod
    public fun sendUnsentReports(call: PluginCall) {
        implementation.sendUnsentReports()
        call.resolve()
    }

    @PluginMethod
    public fun deleteUnsentReports(call: PluginCall) {
        implementation.deleteUnsentReports()
        call.resolve()
    }

    @PluginMethod
    public fun recordException(call: PluginCall) {
        val message = call.getString("message") ?: throw PluginException(ERROR_MESSAGE_MISSING)

        val stacktrace = call.getArray("stacktrace", null)
        implementation.recordException(message, stacktrace)
        call.resolve()
    }

    public companion object {
        public const val ERROR_MESSAGE_MISSING: String = "message must be provided."
        public const val ERROR_KEY_MISSING: String = "key must be provided."
        public const val ERROR_VALUE_MISSING: String = "value must be provided."
        public const val ERROR_USERID_MISSING: String = "userId must be provided."
        public const val ERROR_ENABLED_MISSING: String = "enabled must be provided."
    }
}

File: ios/Sources/FirebaseCrashlyticsPlugin/FirebaseCrashlyticsPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FirebaseCrashlyticsPlugin)
public class FirebaseCrashlyticsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FirebaseCrashlyticsPlugin"
    public let jsName = "FirebaseCrashlytics"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("crash", FirebaseCrashlyticsPlugin.crash),
        .promise("setContext", FirebaseCrashlyticsPlugin.setContext),
        .promise("addLogMessage", FirebaseCrashlyticsPlugin.addLogMessage),
        .promise("setUserId", FirebaseCrashlyticsPlugin.setUserId),
        .promise("setEnabled", FirebaseCrashlyticsPlugin.setEnabled),
        .promise("isEnabled", FirebaseCrashlyticsPlugin.isEnabled),
        .promise("didCrashDuringPreviousExecution", FirebaseCrashlyticsPlugin.didCrashDuringPreviousExecution),
        .promise("sendUnsentReports", FirebaseCrashlyticsPlugin.sendUnsentReports),
        .promise("deleteUnsentReports", FirebaseCrashlyticsPlugin.deleteUnsentReports),
        .promise("recordException", FirebaseCrashlyticsPlugin.recordException)
    ]

    // Every method stays synchronous: the Crashlytics calls are quick, and the bridge queue keeps the keys, logs,
    // settings and recorded exceptions in the order of the calls.

    public let errorMessageMissing = "message must be provided."
    public let errorKeyMissing = "key must be provided."
    public let errorValueMissing = "value must be provided."
    public let errorUserIdMissing = "userId must be provided."
    public let errorEnabledMissing = "enabled must be provided."
    private var implementation: FirebaseCrashlytics?

    override public func load() {
        implementation = FirebaseCrashlytics()
    }

    func crash(_ call: CAPPluginCall) {
        call.resolve()
        implementation?.crash()
    }

    func setContext(_ call: CAPPluginCall) throws {
        guard let key = call.getString("key") else {
            throw CAPPluginError(errorKeyMissing)
        }
        let hasValue = call.options["value"] != nil
        if hasValue == false {
            throw CAPPluginError(errorValueMissing)
        }
        let type = call.getString("type") ?? "string"
        implementation?.setContext(key, type, call)
        call.resolve()
    }

    func addLogMessage(_ call: CAPPluginCall) throws {
        guard let message = call.getString("message") else {
            throw CAPPluginError(errorMessageMissing)
        }
        implementation?.addLogMessage(message)
        call.resolve()
    }

    func setUserId(_ call: CAPPluginCall) throws {
        guard let userId = call.getString("userId") else {
            throw CAPPluginError(errorUserIdMissing)
        }
        implementation?.setUserID(userId)
        call.resolve()
    }

    func setEnabled(_ call: CAPPluginCall) throws {
        guard let enabled = call.getBool("enabled") else {
            throw CAPPluginError(errorEnabledMissing)
        }
        implementation?.setEnabled(enabled)
        call.resolve()
    }

    func isEnabled(_ call: CAPPluginCall) {
        let enabled = implementation?.isEnabled()
        call.resolve([
            "enabled": enabled!
        ])
    }

    func didCrashDuringPreviousExecution(_ call: CAPPluginCall) {
        let crashed = implementation?.didCrashDuringPreviousExecution()
        call.resolve([
            "crashed": crashed!
        ])
    }

    func sendUnsentReports(_ call: CAPPluginCall) {
        implementation?.sendUnsentReports()
        call.resolve()
    }

    func deleteUnsentReports(_ call: CAPPluginCall) {
        implementation?.deleteUnsentReports()
        call.resolve()
    }

    func recordException(_ call: CAPPluginCall) throws {
        guard let message = call.getString("message") else {
            throw CAPPluginError(errorMessageMissing)
        }

        let stacktrace = call.getArray("stacktrace", JSObject.self)
        if stacktrace == nil || stacktrace!.isEmpty {
            let domain = call.getString("domain") ?? ""
            let code = call.getInt("code") ?? -1001

            implementation?.recordException(message, domain, code)
        } else {
            implementation?.recordExceptionWithStacktrace(message, stacktrace!)
        }
        call.resolve()
    }
}
